### PR TITLE
[ROCm] increase timeout, enable test_backend_group

### PR DESCRIPTION
**Pull request by jeffdaily**
CC @iotamudelta

### CODE REVIEW

File: test/distributed/test_distributed.py [hunk 1]
```diff
@@ -247,7 +247,7 @@ def init(cls):
             os.unlink(os.path.join(barrier_dir, f_name))
 
     @classmethod
-    def sync(cls, wait_for=None, timeout=5):
+    def sync(cls, wait_for=None, timeout=10):
```
Comment (pritamdamania87) [inline test/distributed/test_distributed.py:250]: What is the reason for this change?

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]
Looks like there are a bunch of CI failures.

**Comment by jeffdaily** [comment]
Some CI failures look unrelated to the change.
```
Unable to checkout '815e209e4f30910dfdc47aa2ad041e3a46d61b44' in submodule path 'third_party/fbgemm'
```

File: test/distributed/test_distributed.py (hunk 1 above)
Comment (jeffdaily) [reply test/distributed/test_distributed.py:250]: The subtests `TestDistBackend.test_backend_group` and `TestDistBackend.test_backend_full_group`  were recently skipped for the ROCm platform due to ROCm CI failures.  The root cause was a timeout:

```
Process process 0:
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/pytorch/test/distributed/test_distributed.py", line 2212, in _run
    self._barrier()
  File "/pytorch/test/distributed/test_distributed.py", line 280, in _barrier
    Barrier.sync(*args, **kwargs)
  File "/pytorch/test/distributed/test_distributed.py", line 274, in sync
    raise RuntimeError("barrier timeout")
RuntimeError: barrier timeout
FAIL
```

**Comment by jeffdaily** [comment]
clang-format failures are not from files touch by this PR.
```
torch/csrc/jit/codegen/cuda/dispatch.h is not formatted correctly
torch/csrc/jit/codegen/cuda/interface.cpp is not formatted correctly
```

File: test/distributed/test_distributed.py (hunk 1 above)
Comment (pritamdamania87) [reply test/distributed/test_distributed.py:250]: I see, can we run the test in asan mode (usually the slowest) about 100 times locally to make sure the test isn't flaky with the new timeout?

**Comment by pritamdamania87** [comment]
> Some CI failures look unrelated to the change.

Maybe a rebase would fix those failures.

**Comment by jeffdaily** [comment]
@ezyang After rebase, I can't build this PR with latest master due to errors such as 
```
In file included from /pytorch/build/third_party/gloo/hip/gloo/hip_allreduce_ring.cc:9:
/pytorch/build/third_party/gloo/hip/gloo/hip_allreduce_ring.h:26:11: error: exception specification of overriding function is more lax than base version
  virtual ~HipAllreduceRing() = default;
          ^
```
Perhaps this is fixed by https://github.com/facebookincubator/gloo/commit/e8aaa347017f5600ed364a016f26b691ff8aa3fa but gloo submodule isn't updated yet.

**Comment by ezyang** [comment]
ok, then you'll need to do a gloo submodule update first. Submit a PR for it and assign me or someone else as reviewer on it.

File: test/distributed/test_distributed.py (hunk 1 above)
Comment (jeffdaily) [reply test/distributed/test_distributed.py:272]: Is asan mode supported on HIP/CUDA?  Only reference I find to asan mode  is [here](https://github.com/pytorch/pytorch/blob/d6b51e4adfb34a177c89ed966e245abd49b14833/.jenkins/pytorch/build-asan.sh#L36), and it seems `USE_CUDA=0` for asan builds.

**Comment by ezyang** [comment]
ROCm tests are failing on this. Do you want me to go ahead and merge it anyway? I know master is broken righ tnow.

**Comment by jeffdaily** [comment]
Rebased to fix CI tests.
